editor/state.js

```diff
@@ -92,17 +92,49 @@ export const blocks = combineUndoableReducers( {
  * @param  {Object} action Dispatched action
  * @return {Object}        Updated state
  */
-export function selectedBlock( state = null, action ) {
+export function selectedBlock( state = {}, action ) {
 	switch ( action.type ) {
 		case 'TOGGLE_BLOCK_SELECTED':
-			return action.selected ? action.uid : null;
+			if ( ! action.selected ) {
+				return state.uid === action.uid ? {} : state;
+			}
+			return action.uid === state.uid
+				? state
+				: { uid: action.uid, typing: false, focus: {} };
 
 		case 'MOVE_BLOCK_UP':
 		case 'MOVE_BLOCK_DOWN':
-			return action.uid;
+			return action.uid === state.uid
+				? state
+				: { uid: action.uid, typing: false, focus: {} };
 
 		case 'INSERT_BLOCK':
-			return action.block.uid;
+			return {
+				uid: action.block.uid,
+				typing: false,
+				focus: {}
+			};
+
+		case 'UPDATE_FOCUS':
+			return {
+				uid: action.uid,
+				typing: state.uid === action.uid ? state.typing : false,
+				focus: action.config
+			};
+
+		case 'START_TYPING':
+			if ( action.uid !== state.uid ) {
```

Comment:
Like implemented for now "START_TYPING" is per-block. It's triggered when typing inside a specific block, we can drop the `uid`. I feel it should stay as is unless we catch the start typing in a higher level maybe?